compiler/src/dotty/tools/dotc/transform/MacroAnnotations.scala

```diff
@@ -3,124 +3,145 @@ package transform
+            val (transformedTrees, transformedCompanion) = callMacro(macroInterpreter, tree, companion, annot)
+            // Establish the trees order and check the integrity of the trees
+            transformedTrees.span(_.symbol != tree.symbol) match
+              case (newPrefixed, newTree :: newSuffixed) =>
+                // Check the integrity of the generated trees
+                for prefixedTree <- newPrefixed do checkMacroDef(prefixedTree, tree, annot)
+                for suffixedTree <- newSuffixed do checkMacroDef(suffixedTree, tree, annot)
+                for tcompanion <- transformedCompanion do TreeChecker.checkMacroGeneratedTree(companion.get, tcompanion)
+                TreeChecker.checkMacroGeneratedTree(tree, newTree)
+                (prefixed ::: newPrefixed, newTree, transformedCompanion, newSuffixed ::: suffixed)
+              case (Nil, Nil) =>
+                report.error(i"Unexpected `Nil` returned by `(${annot.tree}).transform(..)` during macro expansion", annot.tree.srcPos)
+                lastResult
+              case (_, Nil) =>
+                report.error(i"Transformed tree for ${tree} was not return by `(${annot.tree}).transform(..)` during macro expansion", annot.tree.srcPos)
+                lastResult
+          else
+            lastResult
+      end val
+
+      // Complete the list of transformed/generated definitions
+      val result = prefixed ::: transformedTree :: suffixed
       debug.println(result.map(_.show).mkString("expanded to:\n", "\n", ""))
-      result
+      (result, transformedCompanion)
+  end expandAnnotations
 
-  /** Interpret the code `new annot(..).transform(using <Quotes(ctx)>)(<tree>)` */
-  private def callMacro(interpreter: Interpreter, tree: MemberDef, annot: Annotation)(using Context): List[MemberDef] =
-    // TODO: Remove when scala.annaotaion.MacroAnnotation is no longer experimental
+  /** Interpret the code `new annot(..).transform(using <Quotes(ctx)>)(<tree>, <companion>)` */
+  private def callMacro(interpreter: Interpreter, tree: MemberDef, companion: Option[MemberDef], annot: Annotation)
+                       (using Context): (List[MemberDef], Option[MemberDef]) =
+    // TODO: Remove when scala.annotation.MacroAnnotation is no longer experimental
     import scala.reflect.Selectable.reflectiveSelectable
     type MacroAnnotation = {
-      def transform(using Quotes)(tree: Object/*Erased type of quotes.refelct.Definition*/): List[MemberDef /*quotes.refelct.Definition known to be MemberDef in QuotesImpl*/]
+      def transform(using Quotes)
+                   (tree: Object/*Erased type of quotes.reflect.Definition*/, companion: Option[Object/*Erased type of quotes.reflect.Definition*/])
+                  : List[MemberDef /*quotes.refelct.Definition known to be MemberDef in QuotesImpl*/]
     }
 
     // Interpret macro annotation instantiation `new myAnnot(..)`
+    // TODO: Make this error handling stronger (no error handling at the moment)
```

Comment:
Then we should fix this now. Maybe this would be enough. Just check that the interpretation error has been reported.
```scala
    val annotInstance = interpreter.interpret[MacroAnnotation](annot.tree) match
      case Some(annotInstance) => annotInstance
      case None => return (List(tree), companion)
```

Reply:
We agreed to investigate this later